src/main/scala/nodes/learning/LeastSquaresEstimator.scala

```diff
@@ -0,0 +1,87 @@
+package nodes.learning
+
+import breeze.linalg._
+import nodes.util.{Densify, Sparsify}
+import org.apache.spark.rdd.RDD
+import pipelines.Logging
+import workflow.{Pipeline, WeightedNode, LabelEstimator, OptimizableLabelEstimator}
+
+import scala.reflect._
+
+/**
+ * A least squares solver that is optimized to use a fast algorithm, based on characteristics of the
+ * workload and cost models.
+ *
+ * Currently selects between Dense LBFGS, Sparse LBFGS, Exact Distributed Solve, and Approximate Block Solve.
+ *
+ * The default weights were determined empirically via results run on a 16 r3.4xlarge node cluster.
+ *
+ * @param lambda  The L2 regularization parameter to use, defaults to 0
+ * @param numMachines
+ * @param cpuWeight
+ * @param memWeight
+ * @param networkWeight
+ * @tparam T
+ */
+class LeastSquaresEstimator[T <: Vector[Double]: ClassTag](
+    lambda: Double = 0,
+    numMachines: Option[Int] = None,
+    cpuWeight: Double = 3.8e-4,
+    memWeight: Double = 2.9e-1,
+    networkWeight: Double = 1.32)
+  extends OptimizableLabelEstimator[T, DenseVector[Double], DenseVector[Double]]
+    with WeightedNode
+    with Logging {
+
+  val options: Seq[(SolverCostModel, (RDD[T], RDD[DenseVector[Double]]) => Pipeline[T, DenseVector[Double]])] = Seq(
+    {
+      val solver = new DenseLBFGSwithL2[T](new LeastSquaresDenseGradient, regParam = lambda, numIterations = 20)
+      (solver, solver.withData(_, _))
+    },
+    {
+      val solver = new SparseLBFGSwithL2(new LeastSquaresSparseGradient, regParam = lambda, numIterations = 20)
+      (solver, Sparsify() andThen (solver, _, _))
+    },
+    {
+      val solver = new BlockLeastSquaresEstimator(1000, 3, lambda = lambda)
+      (solver, Densify() andThen (solver, _, _))
+    },
+    {
+      val solver = new LinearMapEstimator(Some(lambda))
+      (solver, Densify() andThen (solver, _, _))
+    }
```

Comment:
When @etrain made the cost models for me I don't think he bothered w/ this one because it's about equal to DenseLBFGS on the VOC workload